### ISSUE TITLE
Broken GDExtention's resource properties

### ISSUE THREAD

**Issue by Withaust** [godotengine/godot-cpp#1042]
### Godot version

4.0.rc2

### System information

Windows 10, Forward Clustered

### Issue description

After RC2 in `GDExtention` properties that are resources (for example, PackedScenes) get reset to \<empty\> no matter what you do (even if you assign manually from `.tscn` file)
![image](https://user-images.githubusercontent.com/118038102/218985317-5aba935a-7527-4628-aaab-134381479775.png)

### Steps to reproduce

Have an exposed resource property with GDExtention and try assigning any resource to it.

### Minimal reproduction project

[](https://github.com/godotengine/godot/files/10741378/_.zip)

**Comment by Withaust** [comment]
After manually reverting lines changed by https://github.com/godotengine/godot-cpp/pull/1037 everything started working again
![image](https://user-images.githubusercontent.com/118038102/219011120-9d561021-4166-4ddb-a240-7a8790b3bbaf.png)

**Comment by dsnopek** [comment]
I am able to reproduce this issue! Here's a minimal test project that I created: 
[resource-property-test.zip](https://github.com/godotengine/godot-cpp/files/10745312/resource-property-test.zip)

I'm still trying to figure out what exactly is happening...

**Comment by dsnopek** [comment]
I'm starting to understand what's going on here:

- The `Variant` containing the `PackedScene` is being passed from the Godot side to the godot-cpp side
- When `Ref<T>::Ref(const Variant &p_variant)` gets it, it calls `Variant::operator Object *()` to get the `godot::Object` (the godot-cpp version of the object - I'll put `godot::` in front to try and make that distinction)
- Inside `Variant::operator Object*()` there is this line at the end:
  ```
  return reinterpret_cast<Object *>(internal::gde_interface->object_get_instance_binding(obj, internal::token, &Object::___binding_callbacks));
  ```
- Via a whole bunch of indirection, this leads to a `godot::Object` being constructed (because the `&Object::__binding_callbacks` are being passed in), rather than a `godot::PackedScene`
- So, the `dynamic_cast<godot::PackedScene>()` ultimately fails, because the object isn't a `godot::PackedScene` (even though the `_owner` property points to a `PackedScene` on the Godot side)

Before PR #1037, it was doing a `reinterpret_cast<T>()` (instead of `dynamic_cast<T>()`) which would cast the `godot::Object` to a `godot::PackedScene`, even though it wasn't actually a `godot::PackedScene` object. I think this would have made calling any of `godot::PackedScene`'s methods unsafe?

So, I think the solution is figuring out how to have `Variant::operator Object*()` (on the godot-cpp side - this is getting confusing :-)) actually create a `godot::PackedScene` rather than a `godot::Object`? I'm not yet sure how to do that, though.

I'd love some review of my reasoning above from someone who is more familiar with how GDExtension is intended to work. :-)

**Comment by dsnopek** [comment]
I think I've got a way to fix this issue but it's kind of involved (requires a change on the Godot engine side, and a new generated file in godot-cpp) and so will probably require a bunch of review and back-and-forth. With the 4.0 stable release coming soon, let's just revert PR #1037 for now, and I can finish a proper fix after 4.0 :-)

Here's a PR #1046 to do the revert
